src/Core/Types.Selection/SingleOrDefaultMiddleware~1.cs

```diff
@@ -0,0 +1,39 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using HotChocolate.Resolvers;
+
+namespace HotChocolate.Types
+{
+    public class SingleOrDefaultMiddleware<T>
+    {
+        private readonly FieldDelegate _next;
+
+        public SingleOrDefaultMiddleware(FieldDelegate next)
+        {
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+        }
+
+        public async Task InvokeAsync(IMiddlewareContext context)
+        {
+            await _next(context).ConfigureAwait(false);
+
+            IQueryable<T> source = null;
+
+            if (context.Result is IQueryable<T> q)
+            {
+                source = q;
+            }
+            else if (context.Result is IEnumerable<T> e)
+            {
+                source = e.AsQueryable();
+            }
+            if (source != null)
+            {
+                context.Result = await Task.Run(
+                    () => source.SingleOrDefault(), context.RequestAborted);
+            }
```

Comment:
Would it also be possible to add support for IAsyncEnumerable, for query providers which offer direct async support, such as Entity Framework Core?  The process of running the queryable inside of a Task.Run call and awaiting it for such providers would end up creating a new thread (or queuing the task to the thread pool), then blocking that thread as it runs the asynchronous query synchronously.  For .NET Standard 2.0 and .NET Framework 4.6.1 (which this library targets), the IAsyncEnumerable is defined in the Microsoft.Bcl.AsyncInterfaces, which is also what Entity Framework Core references for those target frameworks.  Also, if referencing the System.Linq.Async package, which supports targeting .NET Standard 2.0 and .NET Framework 4.6.1, the implementation would be quite simple:
```csharp
if (context.Result is IAsyncEnumerable<T> ae)
{
    context.Result = await ae.SingleOrDefaultAsync(context.RequestAborted);
    return;
}
```

For query providers which do not support IAsyncEnumerable, either they do not perform asynchronous data access, or they do not easily expose it, thus running it in a Task.Run would be a proper implementation.

Reply:
We have reworked it now.